### ISSUE TITLE
thundersvm-predict crashes due to "out of host memory" error

### ISSUE THREAD

**Issue by julianofoleiss** [zeyiwen/thundersvm#53]
thundersvm-predict crashes when #instances or #features is too large.

```
2018-04-24 13:10:01,033 INFO [default] loading dataset from file "../dataset/test_features_exballroom-ds_linspace-fs_40-nfeats_1-f_libsvm.txt"
2018-04-24 13:10:02,081 INFO [default] #instances = 16720, #features = 156
2018-04-24 13:10:02,920 FATAL [default] out of host memory
2018-04-24 13:10:02,920 WARNING [default] Aborting application. Reason: Fatal log at [/home/juliano/thundersvm/src/thundersvm/thundersvm-predict.cpp:68]
```
This is due to the hard-coded value of 10000 batch_size parameter in a call to model->predict in thundersvm-predict.cpp:

```
vector<float_type> predict_y;
predict_y = model->predict(predict_dataset.instances(), 10000);
```

Changing 10000 to 1000 fixes the crash on my GTX780 with 3GB. Maybe letting this as a command-line parameter will help.

In a related problem, the prediction after the training takes a VERY LONG TIME (when executing thundersvm-train + SVC + large training set). When looking at the code for prediction in thunder-svm-train.cpp:

```
        vector<float_type> predict_y;
        if (parser.do_cross_validation) {
            predict_y = model->cross_validation(train_dataset, parser.param_cmd, parser.nr_fold);
        } else {
            model->train(train_dataset, parser.param_cmd);
            LOG(INFO) << "training finished";
            model->save_to_file(parser.model_file_name);
            LOG(INFO) << "evaluating training score";
            predict_y = model->predict(train_dataset.instances(), 100);
        }
```
batch_size for predict is only 100.. I understand why the default is lower here, since the training set is usually much larger than the validation/test set. However, i believe that leaving this as a command-line parameter allows the user to tweak it according to the available GPU memory.

Thanks!

**Comment by zeyiwen** [comment]
Thanks for the feedback! We'll take a look at this issue.

**Comment by shijiashuai** [comment]
I have added `-m` option to constrain memory usage for training and prediction.

**Comment by shijiashuai** [comment]
I have added `-m` option to constrain memory usage for training and prediction.

**Comment by shijiashuai** [comment]
I have added `-m` option to constrain memory usage for training and prediction.
